{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8697a0b-f0d1-4f1f-bb83-0a60ec1daae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import yaml\n",
    "\n",
    "alphabet = 'ACGT'\n",
    "\n",
    "def empirical_frequencies(counts):\n",
    "    '''\n",
    "    convert a nucleotide counts (dictionary {'A':3423, 'C':34543, ...}) to frequency\n",
    "    '''\n",
    "    total = sum(list(counts.values()))\n",
    "    return np.array([counts[n]/total for n in alphabet])\n",
    "\n",
    "\n",
    "def spectrum_to_matrix(spec):\n",
    "    '''\n",
    "    convert dictionary of mutation counts to mutation matrix\n",
    "    '''\n",
    "    M = np.zeros((4,4))\n",
    "    for i1,n1 in enumerate(alphabet):\n",
    "        for i2,n2 in enumerate(alphabet):\n",
    "            if n1!=n2:\n",
    "                M[i2,i1] = spec[f\"{n1}to{n2}\"]\n",
    "    # normalize off-diagonal rates (just for standardization, doesn't affect the results)\n",
    "    M /= M.sum()\n",
    "    # will the diagonal with 'outflow' term to guarantee conservation of probability\n",
    "    d = M.sum(axis=0)\n",
    "    np.fill_diagonal(M,-d)\n",
    "\n",
    "    return M\n",
    "\n",
    "\n",
    "def equilibrium_probabilities(M):\n",
    "    evals, evecs = np.linalg.eig(M)\n",
    "    # find zero eigenvalue\n",
    "    ii = np.argmin(np.abs(evals))\n",
    "    assert np.abs(evals[ii])<1e-10\n",
    "    # pull out corresponding eigenvector, return normalized to sum_i p_i = 1\n",
    "    p = evecs[:,ii]\n",
    "    return p/p.sum()\n",
    "\n",
    "def get_ffsyn(ref):\n",
    "    from Bio import SeqIO\n",
    "    features  = [{'pos':i, 'nuc':n} for i,n in enumerate(ref.seq)]\n",
    "    ## make a map of codon positions, ignore orf9b (overlaps N)\n",
    "    for feat in ref.features:\n",
    "        if feat.type=='CDS':\n",
    "            gene_name = feat.qualifiers['gene'][0] if \"gene\" in feat.qualifiers else None\n",
    "            if gene_name!='ORF9b':\n",
    "                for gpos, pos in enumerate(feat.location):\n",
    "                    tmp = {'pos_in_codon':(gpos%3)+1, 'gene':gene_name, 'four_fold':False}\n",
    "                    if (gpos%3)+1==3:\n",
    "                        codon12 = \"\".join(ref.seq[pos-2:pos])\n",
    "                        if codon12 in ['TC','CT', 'CC', 'CG', 'AC', 'GT', 'GC','GG']:\n",
    "                            tmp['four_fold'] = True\n",
    "                    features[pos].update(tmp)\n",
    "\n",
    "    return pd.DataFrame(features)\n",
    "\n",
    "def get_sequence_from_genbank(accession):\n",
    "    from Bio import SeqIO\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = \"richard.neher@unibas.ch\"\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    return record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a7a0768-3f8d-4395-8d1d-372981e9b622",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20A (B.1)</td>\n",
       "      <td>0.162199</td>\n",
       "      <td>0.072122</td>\n",
       "      <td>0.027598</td>\n",
       "      <td>0.738081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20B (B.1.1)</td>\n",
       "      <td>0.167663</td>\n",
       "      <td>0.069790</td>\n",
       "      <td>0.028297</td>\n",
       "      <td>0.734250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20C (B.1.367)</td>\n",
       "      <td>0.177134</td>\n",
       "      <td>0.069814</td>\n",
       "      <td>0.030583</td>\n",
       "      <td>0.722470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20E (B.1.177)</td>\n",
       "      <td>0.165593</td>\n",
       "      <td>0.066128</td>\n",
       "      <td>0.036844</td>\n",
       "      <td>0.731436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20G (B.1.2)</td>\n",
       "      <td>0.178689</td>\n",
       "      <td>0.063486</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.726599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20I (Alpha)</td>\n",
       "      <td>0.177130</td>\n",
       "      <td>0.063275</td>\n",
       "      <td>0.035565</td>\n",
       "      <td>0.724030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>21C (Epsilon)</td>\n",
       "      <td>0.185620</td>\n",
       "      <td>0.065432</td>\n",
       "      <td>0.034616</td>\n",
       "      <td>0.714331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>21I (Delta)</td>\n",
       "      <td>0.142835</td>\n",
       "      <td>0.080941</td>\n",
       "      <td>0.030449</td>\n",
       "      <td>0.745775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>21J (Delta)</td>\n",
       "      <td>0.144964</td>\n",
       "      <td>0.084301</td>\n",
       "      <td>0.030591</td>\n",
       "      <td>0.740144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>21K (Omicron BA.1)</td>\n",
       "      <td>0.248047</td>\n",
       "      <td>0.075088</td>\n",
       "      <td>0.061554</td>\n",
       "      <td>0.615312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21L (Omicron BA.2)</td>\n",
       "      <td>0.214250</td>\n",
       "      <td>0.070823</td>\n",
       "      <td>0.055136</td>\n",
       "      <td>0.659791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>22A (Omicron BA.4)</td>\n",
       "      <td>0.218087</td>\n",
       "      <td>0.070782</td>\n",
       "      <td>0.047083</td>\n",
       "      <td>0.664048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>22B (Omicron BA.5)</td>\n",
       "      <td>0.232705</td>\n",
       "      <td>0.073487</td>\n",
       "      <td>0.045475</td>\n",
       "      <td>0.648332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>22C (Omicron BA.2.12.1)</td>\n",
       "      <td>0.200377</td>\n",
       "      <td>0.077208</td>\n",
       "      <td>0.050313</td>\n",
       "      <td>0.672101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      virus         A         C         G         T\n",
       "0                 20A (B.1)  0.162199  0.072122  0.027598  0.738081\n",
       "1               20B (B.1.1)  0.167663  0.069790  0.028297  0.734250\n",
       "2             20C (B.1.367)  0.177134  0.069814  0.030583  0.722470\n",
       "3             20E (B.1.177)  0.165593  0.066128  0.036844  0.731436\n",
       "4               20G (B.1.2)  0.178689  0.063486  0.031226  0.726599\n",
       "5               20I (Alpha)  0.177130  0.063275  0.035565  0.724030\n",
       "6             21C (Epsilon)  0.185620  0.065432  0.034616  0.714331\n",
       "7               21I (Delta)  0.142835  0.080941  0.030449  0.745775\n",
       "8               21J (Delta)  0.144964  0.084301  0.030591  0.740144\n",
       "9        21K (Omicron BA.1)  0.248047  0.075088  0.061554  0.615312\n",
       "10       21L (Omicron BA.2)  0.214250  0.070823  0.055136  0.659791\n",
       "11       22A (Omicron BA.4)  0.218087  0.070782  0.047083  0.664048\n",
       "12       22B (Omicron BA.5)  0.232705  0.073487  0.045475  0.648332\n",
       "13  22C (Omicron BA.2.12.1)  0.200377  0.077208  0.050313  0.672101"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if os.path.isfile(\"config.yaml\"):\n",
    "    basedir = \"./\"\n",
    "else:\n",
    "    basedir = \"../\"\n",
    "    \n",
    "with open(os.path.join(basedir, \"config.yaml\")) as f:\n",
    "    config = yaml.safe_load(f)\n",
    "clade_synonyms = config[\"clade_synonyms\"]\n",
    "\n",
    "rates_by_clade =  pd.read_csv(os.path.join(basedir, \"results/synonymous_mut_rates/rates_by_clade.csv\"))\n",
    "\n",
    "clades = rates_by_clade.clade.unique()\n",
    "\n",
    "equilibrium_probabilities_by_clade = {}\n",
    "for clade in clades:\n",
    "    subset = rates_by_clade.loc[rates_by_clade.clade==clade]\n",
    "    tmp_spectrum = {}\n",
    "    for rate in subset.itertuples():\n",
    "        tmp_spectrum[rate.mut_type] = rate.rate\n",
    "\n",
    "    equilibrium_probabilities_by_clade[clade] = equilibrium_probabilities(spectrum_to_matrix(tmp_spectrum))\n",
    "\n",
    "eqprob_df = (\n",
    "    pd.DataFrame.from_dict(\n",
    "        equilibrium_probabilities_by_clade,\n",
    "        orient=\"index\",\n",
    "        columns=list(alphabet)\n",
    "    )\n",
    "    .applymap(lambda x: x.real)\n",
    "    .rename_axis(\"virus\")\n",
    "    .reset_index()\n",
    "    .assign(virus=lambda x: x[\"virus\"] + \" (\" + x[\"virus\"].map(clade_synonyms) + \")\")\n",
    ")\n",
    "\n",
    "eqprob_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9663bd2a-ad02-4653-b233-715b98dff773",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing accession='MN908947'\n",
      "Parsing accession='MN996532.2'\n",
      "Parsing accession='MZ937000.1'\n",
      "Parsing accession='KY352407.1'\n",
      "Parsing accession='AY278488.2 '\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2 (MN908947)</td>\n",
       "      <td>0.289660</td>\n",
       "      <td>0.136922</td>\n",
       "      <td>0.064920</td>\n",
       "      <td>0.508499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RaTG13 (MN996532.2)</td>\n",
       "      <td>0.292148</td>\n",
       "      <td>0.134402</td>\n",
       "      <td>0.063900</td>\n",
       "      <td>0.509550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BANAL-52 (MZ937000.1)</td>\n",
       "      <td>0.288941</td>\n",
       "      <td>0.133176</td>\n",
       "      <td>0.066118</td>\n",
       "      <td>0.511765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BtKY72 (KY352407.1)</td>\n",
       "      <td>0.262702</td>\n",
       "      <td>0.152661</td>\n",
       "      <td>0.092222</td>\n",
       "      <td>0.492415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-1 (AY278488.2 )</td>\n",
       "      <td>0.286147</td>\n",
       "      <td>0.166511</td>\n",
       "      <td>0.089552</td>\n",
       "      <td>0.457789</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      virus         A         C         G         T\n",
       "0     SARS-CoV-2 (MN908947)  0.289660  0.136922  0.064920  0.508499\n",
       "1       RaTG13 (MN996532.2)  0.292148  0.134402  0.063900  0.509550\n",
       "2     BANAL-52 (MZ937000.1)  0.288941  0.133176  0.066118  0.511765\n",
       "3       BtKY72 (KY352407.1)  0.262702  0.152661  0.092222  0.492415\n",
       "4  SARS-CoV-1 (AY278488.2 )  0.286147  0.166511  0.089552  0.457789"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accessions = {\n",
    "    \"MN908947\": \"SARS-CoV-2\",\n",
    "    \"MN996532.2\": \"RaTG13\",\n",
    "    \"MZ937000.1\": \"BANAL-52\",\n",
    "    \"KY352407.1\": \"BtKY72\",\n",
    "    \"AY278488.2 \": \"SARS-CoV-1\",\n",
    "}\n",
    "\n",
    "empirical_frequencies_references = {}\n",
    "for accession in accessions:\n",
    "    print(f\"Parsing {accession=}\")\n",
    "    ref = get_sequence_from_genbank(accession)\n",
    "    ref_genome = get_ffsyn(ref)\n",
    "    empirical_frequencies_references[accession] = empirical_frequencies({i:r for i,r in ref_genome.loc[ref_genome.four_fold & (~ref_genome.four_fold.isna())].nuc.value_counts().iteritems()})\n",
    "\n",
    "empirical_freqs_df = (\n",
    "    pd.DataFrame.from_dict(\n",
    "        empirical_frequencies_references,\n",
    "        orient=\"index\",\n",
    "        columns=list(alphabet)\n",
    "    )\n",
    "    .rename_axis(\"virus\")\n",
    "    .reset_index()\n",
    "    .assign(virus=lambda x: x[\"virus\"].map(accessions) + \" (\" + x[\"virus\"] + \")\")\n",
    ")\n",
    "\n",
    "empirical_freqs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4afa20f-c160-498f-913d-cd303f246bea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-c4b31794aca5404a9dbb359f23292960\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-c4b31794aca5404a9dbb359f23292960\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-c4b31794aca5404a9dbb359f23292960\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-be8b5306c96574b0821aa8a268086c11\"}, \"mark\": {\"type\": \"point\", \"opacity\": 1, \"size\": 60, \"stroke\": null}, \"encoding\": {\"facet\": {\"columns\": 1, \"field\": \"group\", \"title\": null, \"type\": \"nominal\"}, \"fill\": {\"field\": \"nucleotide\", \"type\": \"nominal\"}, \"shape\": {\"field\": \"nucleotide\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"virus\", \"type\": \"nominal\"}, {\"field\": \"group\", \"type\": \"nominal\"}, {\"field\": \"nucleotide\", \"type\": \"nominal\"}, {\"field\": \"frequency\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"frequency\", \"title\": \"nucleotide frequency\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"virus\", \"sort\": [\"SARS-CoV-2 (MN908947)\", \"RaTG13 (MN996532.2)\", \"BANAL-52 (MZ937000.1)\", \"BtKY72 (KY352407.1)\", \"SARS-CoV-1 (AY278488.2 )\"], \"title\": null, \"type\": \"nominal\"}}, \"height\": {\"step\": 15}, \"resolve\": {\"scale\": {\"y\": \"independent\"}}, \"width\": 250, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-be8b5306c96574b0821aa8a268086c11\": [{\"virus\": \"20A (B.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.16219870905994999}, {\"virus\": \"20B (B.1.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.1676632277233155}, {\"virus\": \"20C (B.1.367)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.17713392192592808}, {\"virus\": \"20E (B.1.177)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.16559290828502307}, {\"virus\": \"20G (B.1.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.1786892557726426}, {\"virus\": \"20I (Alpha)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.17713028440143117}, {\"virus\": \"21C (Epsilon)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.1856204623429016}, {\"virus\": \"21I (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.1428350326864429}, {\"virus\": \"21J (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.14496425829101428}, {\"virus\": \"21K (Omicron BA.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.24804700937051996}, {\"virus\": \"21L (Omicron BA.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.21424984890507012}, {\"virus\": \"22A (Omicron BA.4)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.21808689388697314}, {\"virus\": \"22B (Omicron BA.5)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.23270532895657786}, {\"virus\": \"22C (Omicron BA.2.12.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"A\", \"frequency\": 0.2003770515934982}, {\"virus\": \"SARS-CoV-2 (MN908947)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"A\", \"frequency\": 0.2896600566572238}, {\"virus\": \"RaTG13 (MN996532.2)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"A\", \"frequency\": 0.2921480782834237}, {\"virus\": \"BANAL-52 (MZ937000.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"A\", \"frequency\": 0.28894117647058826}, {\"virus\": \"BtKY72 (KY352407.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"A\", \"frequency\": 0.26270166144955454}, {\"virus\": \"SARS-CoV-1 (AY278488.2 )\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"A\", \"frequency\": 0.28614738805970147}, {\"virus\": \"20A (B.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07212197140092884}, {\"virus\": \"20B (B.1.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.06978967496670924}, {\"virus\": \"20C (B.1.367)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.06981371125797321}, {\"virus\": \"20E (B.1.177)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.06612774000495124}, {\"virus\": \"20G (B.1.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.0634858247657651}, {\"virus\": \"20I (Alpha)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.06327506207414552}, {\"virus\": \"21C (Epsilon)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.06543225947727693}, {\"virus\": \"21I (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.0809409387213598}, {\"virus\": \"21J (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.08430082878817428}, {\"virus\": \"21K (Omicron BA.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07508755155470123}, {\"virus\": \"21L (Omicron BA.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07082292814656868}, {\"virus\": \"22A (Omicron BA.4)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07078172978591755}, {\"virus\": \"22B (Omicron BA.5)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07348723813913163}, {\"virus\": \"22C (Omicron BA.2.12.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"C\", \"frequency\": 0.07720837127842367}, {\"virus\": \"SARS-CoV-2 (MN908947)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"C\", \"frequency\": 0.1369216241737488}, {\"virus\": \"RaTG13 (MN996532.2)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"C\", \"frequency\": 0.13440226361707144}, {\"virus\": \"BANAL-52 (MZ937000.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"C\", \"frequency\": 0.13317647058823529}, {\"virus\": \"BtKY72 (KY352407.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"C\", \"frequency\": 0.15266072718516735}, {\"virus\": \"SARS-CoV-1 (AY278488.2 )\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"C\", \"frequency\": 0.16651119402985073}, {\"virus\": \"20A (B.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.027598274113124113}, {\"virus\": \"20B (B.1.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.02829733925589806}, {\"virus\": \"20C (B.1.367)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.03058255068312611}, {\"virus\": \"20E (B.1.177)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.03684373182990693}, {\"virus\": \"20G (B.1.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.031225951013096734}, {\"virus\": \"20I (Alpha)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.035564786822025185}, {\"virus\": \"21C (Epsilon)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.034616339449394844}, {\"virus\": \"21I (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.03044857283075827}, {\"virus\": \"21J (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.030591050720738727}, {\"virus\": \"21K (Omicron BA.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.06155390370699836}, {\"virus\": \"21L (Omicron BA.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.05513584802740937}, {\"virus\": \"22A (Omicron BA.4)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.047083184779474416}, {\"virus\": \"22B (Omicron BA.5)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.04547538224318687}, {\"virus\": \"22C (Omicron BA.2.12.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"G\", \"frequency\": 0.050313180475831526}, {\"virus\": \"SARS-CoV-2 (MN908947)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"G\", \"frequency\": 0.06491973559962229}, {\"virus\": \"RaTG13 (MN996532.2)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"G\", \"frequency\": 0.06390002357934449}, {\"virus\": \"BANAL-52 (MZ937000.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"G\", \"frequency\": 0.06611764705882353}, {\"virus\": \"BtKY72 (KY352407.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"G\", \"frequency\": 0.09222248976643391}, {\"virus\": \"SARS-CoV-1 (AY278488.2 )\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"G\", \"frequency\": 0.08955223880597014}, {\"virus\": \"20A (B.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7380810454259971}, {\"virus\": \"20B (B.1.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7342497580540772}, {\"virus\": \"20C (B.1.367)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7224698161329726}, {\"virus\": \"20E (B.1.177)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7314356198801187}, {\"virus\": \"20G (B.1.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7265989684484955}, {\"virus\": \"20I (Alpha)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7240298667023981}, {\"virus\": \"21C (Epsilon)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7143309387304266}, {\"virus\": \"21I (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7457754557614391}, {\"virus\": \"21J (Delta)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.7401438622000727}, {\"virus\": \"21K (Omicron BA.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.6153115353677804}, {\"virus\": \"21L (Omicron BA.2)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.6597913749209517}, {\"virus\": \"22A (Omicron BA.4)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.6640481915476348}, {\"virus\": \"22B (Omicron BA.5)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.6483320506611036}, {\"virus\": \"22C (Omicron BA.2.12.1)\", \"group\": \"predicted equilibrium frequencies\", \"nucleotide\": \"T\", \"frequency\": 0.6721013966522466}, {\"virus\": \"SARS-CoV-2 (MN908947)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"T\", \"frequency\": 0.5084985835694051}, {\"virus\": \"RaTG13 (MN996532.2)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"T\", \"frequency\": 0.5095496345201603}, {\"virus\": \"BANAL-52 (MZ937000.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"T\", \"frequency\": 0.5117647058823529}, {\"virus\": \"BtKY72 (KY352407.1)\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"T\", \"frequency\": 0.4924151215988442}, {\"virus\": \"SARS-CoV-1 (AY278488.2 )\", \"group\": \"virus frequencies at 4-fold degenerate sites\", \"nucleotide\": \"T\", \"frequency\": 0.45778917910447764}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat(\n",
    "    [\n",
    "        eqprob_df.assign(group=\"predicted equilibrium frequencies\"),\n",
    "        empirical_freqs_df.assign(group=\"virus frequencies at 4-fold degenerate sites\"),\n",
    "    ]\n",
    ").melt(\n",
    "    id_vars=[\"virus\", \"group\"],\n",
    "    var_name=\"nucleotide\",\n",
    "    value_name=\"frequency\",\n",
    ")\n",
    "\n",
    "chart = (\n",
    "    alt.Chart(df)\n",
    "    .encode(\n",
    "        x=alt.X(\"frequency\", title=\"nucleotide frequency\"),\n",
    "        y=alt.Y(\"virus\", title=None, sort=empirical_freqs_df[\"virus\"].unique().tolist()),\n",
    "        shape=alt.Shape(\"nucleotide\"),\n",
    "        fill=alt.Fill(\"nucleotide\"),\n",
    "        facet=alt.Facet(\"group\", columns=1, title=None),\n",
    "        tooltip=df.columns.tolist(),\n",
    "    )\n",
    "    .mark_point(stroke=None, size=60, opacity=1)\n",
    "    .resolve_scale(y=\"independent\")\n",
    "    .properties(width=250, height=alt.Step(15))\n",
    ")\n",
    "\n",
    "outfile = os.path.join(basedir, \"results/equilibrium_freqs/equilibrium_freqs.html\")\n",
    "os.makedirs(os.path.dirname(outfile), exist_ok=True)\n",
    "chart.save(outfile)\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5891807d-f46b-4b42-baf0-8e191085517f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
